client/state/timezones/test/reducer.js

```diff
@@ -0,0 +1,335 @@
+/**
+ * External dependencies
+ */
+import { expect } from 'chai';
+import deepFreeze from 'deep-freeze';
+
+/**
+ * Internal dependencies
+ */
+import { useSandbox } from 'test/helpers/use-sinon';
+
+import timezonesReducer, {
+	byContinentsReducer,
+	labelsReducer,
+	rawOffsetReducer,
+} from '../reducer';
+
+import { timezonesReceive } from '../actions';
+import { fromApi } from 'state/data-layer/wpcom/timezones';
+
+/**
+ * Fixture data
+ */
+import {
+	RAW_OFFSETS,
+	TIMEZONES,
+	TIMEZONES_BY_CONTINENT,
+} from './fixture';
+
+describe( 'reducer', () => {
+	let sandbox;
+
+	useSandbox( newSandbox => {
+		sandbox = newSandbox;
+		sandbox.stub( console, 'warn' );
+	} );
+
+	it( 'should export expected reducer keys', () => {
+		expect( timezonesReducer( undefined, {} ) ).to.have.keys( [
+			'byContinents',
+			'labels',
+			'rawOffsets',
+		] );
+	} );
+
+	describe( '#rawOffsetReducer()', () => {
+		it( 'should default to an empty Object', () => {
+			expect( rawOffsetReducer( undefined, {} ) ).to.eql( {} );
+		} );
+
+		it( 'should index timezones.rawOffsets state', () => {
+			const initialState = undefined;
+			const restAPIResponse = {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			};
+
+			const apiResponse = fromApi( restAPIResponse );
+			const action = timezonesReceive( apiResponse );
+
+			const expectedState = apiResponse.rawOffsets;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
+
+		it( 'should override timezones.rawOffsets state', () => {
+			const initialState = fromApi( {
+				manual_utc_offsets: [
+					{ value: 'foo', label: 'bar' }
+				],
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			} ).rawOffsets;
+
+			deepFreeze( initialState );
+
+			const restAPIResponse = {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			};
+
+			const apiResponse = fromApi( restAPIResponse );  // data-layer processing
+			const action = timezonesReceive( apiResponse );
+
+			const expectedState = apiResponse.rawOffsets;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
+
+		it( 'should persist state', () => {
+			const initialState = fromApi( {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			} );
+			deepFreeze( initialState );
+
+			const action = { type: 'SERIALIZE' };
+
+			const expectedState = initialState;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
```

Comment:
> if the schema itself defines the structure then we don't need to test it because JSON schema has its own testing and validation.

It does have it's own testing and validation, but in this case we're not testing that the JSON schema spec works in general. We're testing that we wrote down the schema we wanted, that the schema actually represents the data contract we want to have. Our mock data in those tests become the source of truth, and our tests will be green or red in a deterministic way. So really this is still trying to catch a programming error (misconfiguring a schema), rather than say detect when our api changes unexpectedly or when systems goes down.

Can the mocks get out of date? Yes, but if we update what we store locally, we also need to update our schema and tests. That's our only safeguard that we aren't using old data that doesn't match the shape we expect. 

I think there is value to writing these tests, even if the author throws them away later. Most folks are rather unfamiliar with the JSON schema spec, and tests provide a quick feedback loop to see that they did this correctly.

> we don't have an integration suite to my knowledge.

We have a selenium test runner that performs a number of smoke tests at different viewport sizes for common flows. The tests currently run during every master branch merge.  It's quite handy as a sanity check, and is very useful to test components that regress often, even the ones covered by a number of unit tests.

The [e2e integration](https://github.com/Automattic/wp-e2e-tests) tests are also non-trivial to maintain and are by their nature much more fragile. I would really recommend looking into the project if you're interested in contributing or need to add a smoke test. Any flow patrol folks will be happy to help you troubleshoot setup. 

That said I agree that it may be useful to have the e2e tests complain when we see a console warning from our persistence layer.